### ISSUE TITLE
Can't add a draggable button in column headers

### ISSUE THREAD

**Issue by snowystinger** [adobe/react-spectrum#6195]
### Discussed in https://github.com/adobe/react-spectrum/discussions/6191

<div type='discussions-op-text'>

<sup>Originally posted by **bu3alwa** April 12, 2024</sup>
I made a sandbox and issue is when you add the drag button in the column it does not trigger on the column. Is this a bug? I understand this feature that is not provided but I would think I would be able to extend columns to implement my own features.

https://codesandbox.io/p/live/d531fedb-93b3-4612-90ae-f8c6009e2aaf</div>

### Possible fix

It looks like this usePress is the culprit https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/table/src/useTableColumnHeader.ts#L52
It calls some preventDefaults/stopPropagations which are likely interferring.

I have a vague memory of not applying the returned value from usePress when it should be disabled, but, that caused a different problem. Someone will have to try that out and see if it causes any issues.

**Comment by sookmax** [comment]
Sorry, but it looks like the sandbox link is broken?

<img width="910" alt="Screenshot 2024-04-12 at 11 59 33 PM" src="https://github.com/adobe/react-spectrum/assets/71210554/2652fb66-813f-49e5-a6ec-35d988b13cff">

**Comment by bu3alwa** [comment]
I don't understand codesandbox and why the live link got removed. https://codesandbox.io/p/live/e9b00b28-bfb5-4ec6-a8ff-67b41350ef55 hopefully this is working

**Comment by sookmax** [comment]
> I don't understand codesandbox and why the live link got removed. https://codesandbox.io/p/devbox/laughing-mirzakhani-25fyt7 hopefully this is working

Sorry but this doesn't work either, giving me 'Devbox not found' page..

**Comment by bu3alwa** [comment]
> > I don't understand codesandbox and why the live link got removed. https://codesandbox.io/p/devbox/laughing-mirzakhani-25fyt7 hopefully this is working
> 
> Sorry but this doesn't work either, giving me 'Devbox not found' page..

Sorry, try again now. I don't know why I am having so much trouble with this. Should be set to public now.

**Comment by bu3alwa** [comment]
> I have a vague memory of not applying the returned value from usePress when it should be disabled, but, that caused a different problem. Someone will have to try that out and see if it causes any issues.

So I am just wondering. When you remove those props from usePress on that useTableColumnHeader then sorting won't be a feature that can be used correct? Would it be better to have a seperate hook `useTableColumnSortButton`  and have that implement a button in the header?

**Comment by viktorrenkema** [comment]
@snowystinger were you able to get this to work? I'm having issues with allowing the Table to have both reorderable rows and also reorderable columns, so I'm interested to see how your setup for reordering columns was done.

**Comment by bu3alwa** [comment]
```
diff --git a/node_modules/@react-aria/table/dist/import.mjs b/node_modules/@react-aria/table/dist/import.mjs
index f5bd998..0d3ff62 100644
--- a/node_modules/@react-aria/table/dist/import.mjs
+++ b/node_modules/@react-aria/table/dist/import.mjs
@@ -809,8 +809,9 @@ function $f329116d8ad0aba0$export$9514819a8c81e960(props, state, ref) {
     let isSelectionCellDisabled = node.props.isSelectionCell && state.selectionManager.selectionMode === "single";
     let { pressProps: pressProps } = (0, $lJcFS$usePress)({
         isDisabled: !allowsSorting || isSelectionCellDisabled,
-        onPress () {
+        onPress (e) {
             state.sort(node.key);
+            e.continuePropagation();
         },
         ref: ref
     });
@@ -829,20 +830,21 @@ function $f329116d8ad0aba0$export$9514819a8c81e960(props, state, ref) {
         if (isSortedColumn && sortDirection && (0, $lJcFS$isAndroid)()) sortDescription = `${sortDescription}, ${stringFormatter.format(sortDirection)}`;
     }
     let descriptionProps = (0, $lJcFS$useDescription)(sortDescription);
-    let shouldDisableFocus = state.collection.size === 0;
-    (0, $lJcFS$useEffect)(()=>{
-        if (shouldDisableFocus && state.selectionManager.focusedKey === node.key) state.selectionManager.setFocusedKey(null);
-    }, [
-        shouldDisableFocus,
-        state.selectionManager,
-        node.key
-    ]);
+    // let shouldDisableFocus = state.collection.size === 0;
+    // (0, $lJcFS$useEffect)(()=>{
+    //     if (shouldDisableFocus && state.selectionManager.focusedKey === node.key) state.selectionManager.setFocusedKey(null);
+    // }, [
+    //     shouldDisableFocus,
+    //     state.selectionManager,
+    //     node.key
+    // ]);
     return {
         columnHeaderProps: {
-            ...(0, $lJcFS$mergeProps)(gridCellProps, pressProps, focusableProps, descriptionProps, // If the table is empty, make all column headers untabbable
-            shouldDisableFocus && {
-                tabIndex: -1
-            }),
+            ...(0, $lJcFS$mergeProps)(gridCellProps, focusableProps, descriptionProps, // If the table is empty, make all column headers untabbable
+            // shouldDisableFocus && {
+            //     tabIndex: -1
+            //}
+            ),
             role: "columnheader",
             id: (0, $2140fb2337097f2d$export$37cd4213f2ad742e)(state, node.key),
             "aria-colspan": node.colspan && node.colspan > 1 ? node.colspan : null,
diff --git a/node_modules/@react-aria/table/dist/main.js b/node_modules/@react-aria/table/dist/main.js
index 7f192ed..3aff92b 100644
--- a/node_modules/@react-aria/table/dist/main.js
+++ b/node_modules/@react-aria/table/dist/main.js
@@ -823,8 +823,9 @@ function $7669c34a63ef0113$export$9514819a8c81e960(props, state, ref) {
     let isSelectionCellDisabled = node.props.isSelectionCell && state.selectionManager.selectionMode === "single";
     let { pressProps: pressProps } = (0, $kJQf8$reactariainteractions.usePress)({
         isDisabled: !allowsSorting || isSelectionCellDisabled,
-        onPress () {
+        onPress (e) {
             state.sort(node.key);
+            e.continuePropagation();
         },
         ref: ref
     });
@@ -842,21 +843,18 @@ function $7669c34a63ef0113$export$9514819a8c81e960(props, state, ref) {
         // Android Talkback doesn't support aria-sort so we add sort order details to the aria-described by here
         if (isSortedColumn && sortDirection && (0, $kJQf8$reactariautils.isAndroid)()) sortDescription = `${sortDescription}, ${stringFormatter.format(sortDirection)}`;
     }
-    let descriptionProps = (0, $kJQf8$reactariautils.useDescription)(sortDescription);
-    let shouldDisableFocus = state.collection.size === 0;
-    (0, $kJQf8$react.useEffect)(()=>{
-        if (shouldDisableFocus && state.selectionManager.focusedKey === node.key) state.selectionManager.setFocusedKey(null);
-    }, [
-        shouldDisableFocus,
-        state.selectionManager,
-        node.key
-    ]);
+    // let descriptionProps = (0, $kJQf8$reactariautils.useDescription)(sortDescription);
+    // let shouldDisableFocus = state.collection.size === 0;
+    // (0, $kJQf8$react.useEffect)(()=>{
+    //     if (shouldDisableFocus && state.selectionManager.focusedKey === node.key) state.selectionManager.setFocusedKey(null);
+    // }, [
+    //     shouldDisableFocus,
+    //     state.selectionManager,
+    //     node.key
+    // ]);
     return {
         columnHeaderProps: {
-            ...(0, $kJQf8$reactariautils.mergeProps)(gridCellProps, pressProps, focusableProps, descriptionProps, // If the table is empty, make all column headers untabbable
-            shouldDisableFocus && {
-                tabIndex: -1
-            }),
+            ...(0, $kJQf8$reactariautils.mergeProps)(gridCellProps, focusableProps, descriptionProps), // If the table is empty, make all column headers untabbable
             role: "columnheader",
             id: (0, $6acf696f746f932c$export$37cd4213f2ad742e)(state, node.key),
             "aria-colspan": node.colspan && node.colspan > 1 ? node.colspan : null,
diff --git a/node_modules/@react-aria/table/dist/module.js b/node_modules/@react-aria/table/dist/module.js
index f5bd998..b499df1 100644
--- a/node_modules/@react-aria/table/dist/module.js
+++ b/node_modules/@react-aria/table/dist/module.js
@@ -809,8 +809,9 @@ function $f329116d8ad0aba0$export$9514819a8c81e960(props, state, ref) {
     let isSelectionCellDisabled = node.props.isSelectionCell && state.selectionManager.selectionMode === "single";
     let { pressProps: pressProps } = (0, $lJcFS$usePress)({
         isDisabled: !allowsSorting || isSelectionCellDisabled,
-        onPress () {
+        onPress (e) {
             state.sort(node.key);
+            e.continuePropagation();
         },
         ref: ref
     });
@@ -828,21 +829,18 @@ function $f329116d8ad0aba0$export$9514819a8c81e960(props, state, ref) {
         // Android Talkback doesn't support aria-sort so we add sort order details to the aria-described by here
         if (isSortedColumn && sortDirection && (0, $lJcFS$isAndroid)()) sortDescription = `${sortDescription}, ${stringFormatter.format(sortDirection)}`;
     }
-    let descriptionProps = (0, $lJcFS$useDescription)(sortDescription);
-    let shouldDisableFocus = state.collection.size === 0;
-    (0, $lJcFS$useEffect)(()=>{
-        if (shouldDisableFocus && state.selectionManager.focusedKey === node.key) state.selectionManager.setFocusedKey(null);
-    }, [
-        shouldDisableFocus,
-        state.selectionManager,
-        node.key
-    ]);
+    // let descriptionProps = (0, $lJcFS$useDescription)(sortDescription);
+    // let shouldDisableFocus = state.collection.size === 0;
+    // (0, $lJcFS$useEffect)(()=>{
+    //     if (shouldDisableFocus && state.selectionManager.focusedKey === node.key) state.selectionManager.setFocusedKey(null);
+    // }, [
+    //     shouldDisableFocus,
+    //     state.selectionManager,
+    //     node.key
+    // ]);
     return {
         columnHeaderProps: {
-            ...(0, $lJcFS$mergeProps)(gridCellProps, pressProps, focusableProps, descriptionProps, // If the table is empty, make all column headers untabbable
-            shouldDisableFocus && {
-                tabIndex: -1
-            }),
+            ...(0, $lJcFS$mergeProps)(gridCellProps, focusableProps, descriptionProps), // If the table is empty, make all column headers untabbable
             role: "columnheader",
             id: (0, $2140fb2337097f2d$export$37cd4213f2ad742e)(state, node.key),
             "aria-colspan": node.colspan && node.colspan > 1 ? node.colspan : null,
```

I had to create a patch for the package. I do not remember al the details this would have a side effect and I had to implement something with the column. I hope they address this issue because I am stuck on this version for now.

**Comment by snowystinger** [comment]
> [@snowystinger](https://github.com/snowystinger) were you able to get this to work? I'm having issues with allowing the Table to have both reorderable rows and also reorderable columns, so I'm interested to see how your setup for reordering columns was done.

I've not done reorderable columns. If I did, I would probably create a dialog that I can open from a Column menu which had a ListBox that was reoderable https://react-spectrum.adobe.com/react-aria/ListBox.html#reorderable and that would control the order that the columns are rendered in. This would be accessible and is already possible. You can reference https://react-spectrum.adobe.com/react-aria/Table.html#column-header-menu for how to make a column header menu.

Rows should be pretty easy to do as reorderable, we have an example on the docs as well. https://react-spectrum.adobe.com/react-aria/Table.html#reorderable

The original notes on the issue say that preventDefault from usePress was potentially interfering, I would try again on a nightly release version to see if that is fixed, we just did a major rework of usePress and it fixed a lot of bugs.

**Comment by viktorrenkema** [comment]
> > [@snowystinger](https://github.com/snowystinger) were you able to get this to work? I'm having issues with allowing the Table to have both reorderable rows and also reorderable columns, so I'm interested to see how your setup for reordering columns was done.
> 
> I've not done reorderable columns. If I did, I would probably create a dialog that I can open from a Column menu which had a ListBox that was reoderable https://react-spectrum.adobe.com/react-aria/ListBox.html#reorderable and that would control the order that the columns are rendered in. This would be accessible and is already possible. You can reference https://react-spectrum.adobe.com/react-aria/Table.html#column-header-menu for how to make a column header menu.
> 
> Rows should be pretty easy to do as reorderable, we have an example on the docs as well. https://react-spectrum.adobe.com/react-aria/Table.html#reorderable
> 
> The original notes on the issue say that preventDefault from usePress was potentially interfering, I would try again on a nightly release version to see if that is fixed, we just did a major rework of usePress and it fixed a lot of bugs.

Thanks for the prompt response! Yeah I was hoping that ordering of columns would be possible in the same way as reordering rows. I'm tbh a bit surprised it isn't, all react-aria components are so impressively thorough offering support for far more rare and nuanced interactions than the very common need of reordering columns.

Any idea if there's a reason for this or a technical hurdle or smth?

**Comment by snowystinger** [comment]
A couple reasons, the simplist is that we haven't had time to implement it. Though the reason for that is also tied to it being a difficult problem to solve for accessibility. The way I offered will meet all the accessibility needs today. To do drag and drop for them, we'd need a way for users to start a drag, then we'd need drop targets, then there's announcement handling, etc.

I don't know if we can funnel all that through a drag handle like we do for rows or if we'd need to have a menu with an action to start the drag. If the latter, then we'd need to be able to programmatically support starting a drag, which we also haven't had a chance to implement yet.

Thanks for understanding!

**Comment by viktorrenkema** [comment]
Np, thanks to you for providing context and for all the great work :)

**Comment by snowystinger** [comment]
I'm no longer able to access any of the codesandboxes. So I cannot verify anything. I'm going to close for now since it's your discussion/issue.

If there is still an issue, we can reopen with a new codesandbox.

Thank you for the reminder!
